commons-numbers-complex-arrays/src/main/java/org/apache/commons/numbers/complex/arrays/ComplexList.java

```diff
@@ -0,0 +1,357 @@
+     *
+     * @param index – Index at which the specified element is to be inserted.
+     * @param element – Complex element to be inserted.
+     * @throws IndexOutOfBoundsException – if index isn't within the range of list.
+     */
+    @Override
+    public void add(int index, Complex element) {
+        rangeCheckForInsert(index);
+        double[] e = realAndImagParts;
+        if (size == e.length >>> 1) {
+            e = ensureCapacityInternal(size + 1);
+        }
+        final int i = index << 1;
+        System.arraycopy(e, 2 * index, e, i + 2, (size * 2) - i);
+        e[i] = element.real();
+        e[i + 1] = element.imag();
+        size++;
+    }
+
+    /**
+     * Appends all the elements in the specified collection to the end of this list, in the order that they are
+     * returned by the specified collection's Iterator. The behavior of this operation is undefined if the
+     * specified collection is modified while the operation is in progress.
+     * (This implies that the behavior of this call is undefined if the specified collection is this list,
+     * and this list is nonempty.)
+     *
+     * @param c – Collection containing elements to be added to this list.
+     * @return true if this list changed as a result of the call.
+     * @throws NullPointerException – if the specified collection is null.
+     */
+    @Override
+    public boolean addAll(Collection<? extends Complex> c) {
+        final int numNew = c.size();
+        expand(numNew * 2);
+        double[] realAndImgData = new double[c.size() * 2];
+        int i = 0;
+        for (final Complex val : c) {
+            final int i2 = i << 1;
+            realAndImgData[i2] = val.getReal();
+            realAndImgData[i2 + 1] = val.getImaginary();
+            i++;
+        }
+        System.arraycopy(realAndImgData, 0, realAndImagParts, size * 2, realAndImgData.length);
+        size += numNew;
+        return numNew != 0;
+    }
+
+    /**
+     * Inserts all the elements in the specified collection into this list, starting at the specified position.
+     * Shifts the element currently at that position (if any) and any subsequent elements to the right (increases their indices).
+     * The new elements will appear in the list in the order that they are returned by the specified collection's iterator.
+     *
+     * @param index – Index at which to insert the first element from the specified collection.
+     * @param c – Collection containing elements to be added to this list.
+     * @return true if this list changed as a result of the call.
+     * @throws IndexOutOfBoundsException – if index isn't within the range of list.
+     * @throws NullPointerException – if the specified collection is null.
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends Complex> c) {
+        rangeCheckForInsert(index);
+        final int numNew = c.size();
+        final int numNew2 = numNew << 1;
+        expand(numNew * 2);
```

Comment:
The `expand` method is expecting a number of elements, not a number of doubles. So this should not be doubled.